clarity/contracts/nft-marketplace.clar

```diff
@@ -0,0 +1,237 @@
+)
+  (let ((listing-id (var-get listing-nonce)))
+    ;; Verify that the contract of this asset is whitelisted
+    (asserts! (is-whitelisted (contract-of nft-asset-contract)) ERR_ASSET_CONTRACT_NOT_WHITELISTED)
+    ;; Verify that the asset is not expired
+    (asserts! (> (get expiry nft-asset) burn-block-height) ERR_EXPIRY_IN_PAST)
+    ;; Verify that the asset price is greater than zero
+    (asserts! (> (get price nft-asset) u0) ERR_PRICE_ZERO)
+    ;; Verify that the contract of the payment is whitelisted
+    (asserts! (match (get payment-asset-contract nft-asset)
+      payment-asset
+      (is-whitelisted payment-asset)
+      true
+    ) ERR_PAYMENT_CONTRACT_NOT_WHITELISTED)
+    ;; Transfer the NFT ownership to this contract's principal
+    (try! (transfer-nft
+      nft-asset-contract
+      (get token-id nft-asset)
+      tx-sender
+      (as-contract tx-sender)
+    ))
+    ;; List the NFT in the listings map
+    (map-set listings listing-id (merge
+      { maker: tx-sender, nft-asset-contract: (contract-of nft-asset-contract) }
+      nft-asset
+    ))
+    ;; Increment the nonce to use for the next unique listing ID
+    (var-set listing-nonce (+ listing-id u1))
+    ;; Return the created listing ID
+    (ok listing-id)
+  )
+)
+
+;; Public read-only function to retrieve a listing by its ID
+(define-read-only (get-listing (listing-id uint))
+  (map-get? listings listing-id)
+)
+
+;; Public function to cancel a listing using an asset contract.
+;; This function can only be called by the NFT's creator, and must use the same asset contract that
+;; the NFT uses.
+(define-public (cancel-listing (listing-id uint) (nft-asset-contract <nft-trait>))
+  (let (
+    (listing (unwrap! (map-get? listings listing-id) ERR_UNKNOWN_LISTING))
+    (maker (get maker listing))
+  )
+    ;; Verify that the caller of the function is the creator of the NFT to be cancelled
+    (asserts! (is-eq maker tx-sender) ERR_UNAUTHORISED)
+    ;; Verify that the asset contract to use is the same one that the NFT uses
+    (asserts! (is-eq
+      (get nft-asset-contract listing)
+      (contract-of nft-asset-contract)
+    ) ERR_NFT_ASSET_MISMATCH)
+    ;; Delete the listing
+    (map-delete listings listing-id)
+    ;; Transfer the NFT from this contract's principal back to the creator's principal
+    (as-contract (transfer-nft nft-asset-contract (get token-id listing) tx-sender maker))
+  )
+)
+
+;; Private function to validate that a purchase can be fulfilled
+(define-private (assert-can-fulfil
+  (nft-asset-contract principal)
+  (payment-asset-contract (optional principal))
+  (listing {
+    maker: principal,
+    taker: (optional principal),
+    token-id: uint,
+    nft-asset-contract: principal,
+    expiry: uint,
+    price: uint,
+    payment-asset-contract: (optional principal)
+  })
+)
+  (begin
+    ;; Verify that the buyer is not the same as the NFT creator
+    ;; (asserts! (not (is-eq (get maker listing) tx-sender)) ERR_MAKER_TAKER_EQUAL)
```

Comment:
Do we want to keep this commented? Seems like it might have been commented for testing purposes

Reply:
good catch! I will uncomment it